internal/introspection/idlfile.go

```diff
@@ -0,0 +1,176 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package introspection
+
+import "strings"
+
+// IDLFile is a generic IDL file. For example, a thrift or protobuf IDL.
+type IDLFile struct {
+	// FilePath starts from an arbitrary root containing the file and all its
+	// includes.
+	FilePath string `json:"filePath"`
+
+	// Checksum of Content. The format is: "${type}:${hex}". For example, a
+	// sha1 checksum would look like:
+	// "sha1:d55f7ce6138889e9e45b62df3741820777c65fcd".
+	Checksum string `json:"checksum"`
+
+	// Includes are themselves IDLFile(s).
+	Includes []IDLFile `json:"includes,omitempty"`
+
+	// Content is the IDL file content.
+	Content string `json:"content,omitempty"`
+}
+
+// Flatmap iterates recursively over the idl module and its includes. The cb
+// function is called onces for every unique idl module (based on FilePath).
+// Iteration is aborted if cb returns true.
+func (im *IDLFile) Flatmap(cb func(*IDLFile) bool) {
+	ft := newIDLFlattener(cb)
+	ft.Collect(im)
+}
+
+// IDLFiles is a sortable slice of IDLFile.
+type IDLFiles []IDLFile
+
+func (ims IDLFiles) Len() int {
+	return len(ims)
+}
+
+func (ims IDLFiles) Less(i int, j int) bool {
+	return ims[i].FilePath < ims[j].FilePath
+}
+
+func (ims IDLFiles) Swap(i int, j int) {
+	ims[i], ims[j] = ims[j], ims[i]
+}
+
+// IDLTree builds an IDL tree from a slice of Module.
+func (ims IDLFiles) IDLTree() IDLTree {
+	tb := newIDLTreeBuilder()
+	for i := range ims {
+		tb.Collect(&ims[i])
+	}
+	return tb.Root
+}
+
+// Flatmap iterates recursively over every idl modules. The cb function is
+// called onces for every unique idl module (based on FilePath). Iteration is
+// aborted if cb returns true.
+func (ims IDLFiles) Flatmap(cb func(*IDLFile) bool) {
+	ft := newIDLFlattener(cb)
+	for i := range ims {
+		ft.Collect(&ims[i])
+	}
+}
+
+// IDLTree is a tree of IDLs. A tree can have directories. Dir maps directory
+// names to IDLtrees. And of course, a tree can have any number of idl modules.
+type IDLTree struct {
+	Dir     map[string]*IDLTree `json:"dir,omitempty"`
+	Modules IDLFiles            `json:"modules,omitempty"`
+}
+
+// Compact replaces any tree with a single directory and no IDLFiles by it's
+// single child. The the lone directory name is appended to the parent's
+// directory. I would write an example, but go doc would likely fuck up
+// rendering it anyway.
+func (it *IDLTree) Compact() {
+	for _, l1tree := range it.Dir {
+		l1tree.Compact()
+	}
+	for l1dir, l1tree := range it.Dir {
+		if len(l1tree.Dir) == 1 && len(l1tree.Modules) == 0 {
+			for l2dir, l2tree := range l1tree.Dir {
+				compactedDir := l1dir + "/" + l2dir
+				it.Dir = map[string]*IDLTree{compactedDir: l2tree}
+				break
+			}
+		}
+	}
+}
+
+type idlFlattener struct {
+	seen map[string]struct{}
```

Comment:
idlFlattener, the subject is an IDL. thus its about an idl file. I will rename into `seenIDLs`.